pkg/engine/task/pod_exec.go

```diff
@@ -0,0 +1,232 @@
+	// defined in the PodExec, e.g. when downloading a file, pe.Out will  contain the file bytes.
+	return exec.Stream(so)
+}
+
+// FileSize fetches the size of a file in a remote pod. It runs `stat -c %s file` command in the
+// pod and parses the output.
+func FileSize(file string, pod *v1.Pod, restCfg *rest.Config) (int64, error) {
+	readout, stdout := io.Pipe()
+
+	pe := PodExec{
+		RestCfg:       restCfg,
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"stat", "-c", "%s", file},
+		In:            nil,
+		Out:           stdout,
+		Err:           nil,
+		TTY:           true, // this will forward 2>&1. otherwise, reading from Out will never return for e.g. missing files
+	}
+
+	errCh := make(chan error, 1)
+	go func() {
+		defer stdout.Close() // Close never returns an error
+		errCh <- pe.Run()
+	}()
+
+	buf, err := ioutil.ReadAll(readout)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get the size of %s `: %v", file, err)
+	}
+
+	raw := string(buf)
+
+	// THIS CHECK HAS TO HAPPEN AFTER WE CONSUME THE STDOUT/STDERR READER.
+	if err := <-errCh; err != nil {
+		return 0, fmt.Errorf("failed to get the size of %s, err: %v, %s", file, err, raw)
+	}
+
+	raw = raw[:len(raw)-2] // remove trailing \n\r
+	size, err := strconv.ParseInt(raw, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse the size '%s' of the file %s : %v", raw, file, err)
+	}
+
+	return size, nil
+}
+
+// DownloadFile fetches a file from a remote pod. It runs `tar cf - file` command and streams contents
+// of the file via the stdout. Locally, the tar file is extracted into the passed afero filesystem where
+// it is saved under the same path. Afero filesystem is used to allow the caller downloading and persisting
+// of multiple files concurrently (afero filesystem is thread-safe).
+func DownloadFile(fs afero.Fs, file string, pod *v1.Pod, restCfg *rest.Config) error {
+	readout, writeout := io.Pipe()
+	stderr := bytes.Buffer{}
+
+	pe := PodExec{
+		RestCfg:       restCfg,
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"tar", "cf", "-", file},
+		In:            nil,
+		Out:           writeout,
+		Err:           &stderr,
+	}
+
+	// IMPORTANT:
+	// When downloading files using remotecommand.Executor, we pump the contents of the tar file through
+	// the stdout. PodExec.Run() calls the stream executor, which first writes stdout and stderr of the
+	// command and then returns with an exit code. Since we're using io.Pipe reader and writer which is
+	// tread-safe but SYNCHRONOUS, Run() call will not return until we've consumed the streams (and will
+	// block the execution). For more details on how the underlying streams are copied see remotecommand/v4.go:54
+	//
+	// TL;DR:
+	//  - execute PodExec.Run() in a goroutine when using io.Pipe for Out or Err streams, they
```

Comment:
should this be in the docs for the Run method as well? I wonder what are the best practices here - Run is a public method, if it's implementation is so tricky to use because of some internal details/implementation I wonder if the method itself should not somehow prevent callers from doing things wrong. Because from the interface of the method it actually looks like a pretty innocent method just returning an error :)

Reply:
It is true, that the usage of the `Run()` is somewhat tricky, at least in combination with `io.Pipe` for `Out` and `Err`. Which is what we do for downloading files. This is not necessarily true for other usages but I will a lengthy comment to it.